creator-node/src/services/stateMachineManager/monitoring/StateMonitoringQueue.js

```diff
@@ -0,0 +1,254 @@
+const { getLatestUserId } = require('./utils')
+const processStateMonitoringJob = require('./processStateMonitoringJob')
+
+/**
+ * Handles setup and lifecycle management (producing/adding jobs and consuming/ processing jobs)
+ * of the queue that calculates required syncs and replica set updates (handles user slicing, clocks,
+ * gathering sync metrics, and computing healthy/unhealthy peers).
+ */
+class StateMonitoringQueue {
+  constructor() {
+    this.queue = this.makeQueue(
+      config.get('redisHost'),
+      config.get('redisPort')
+    )
+    this.registerQueueEventHandlers(this.queue)
+    this.registerQueueConsumer(this.queue)
+  }
+
+  async init(audiusLibs) {
+    await this.startQueue(
+      this.queue,
+      audiusLibs.discoveryProvider.discoveryProviderEndpoint,
+      config.get('snapbackModuloBase')
+    )
+  }
+
+  logDebug(msg) {
+    logger.debug(`StateMonitoringQueue DEBUG: ${msg}`)
+  }
+
+  log(msg) {
+    logger.info(`StateMonitoringQueue: ${msg}`)
+  }
+
+  logWarn(msg) {
+    logger.warn(`StateMonitoringQueue WARNING: ${msg}`)
+  }
+
+  logError(msg) {
+    logger.error(`StateMonitoringQueue ERROR: ${msg}`)
+  }
+
+  makeQueue(redisHost, redisPort) {
+    // Settings config from https://github.com/OptimalBits/bull/blob/develop/REFERENCE.md#advanced-settings
+    return new BullQueue(STATE_MONITORING_QUEUE_NAME, {
+      redis: {
+        host: redisHost,
+        port: redisPort
+      },
+      defaultJobOptions: {
+        removeOnComplete: QUEUE_HISTORY,
+        removeOnFail: QUEUE_HISTORY
+      },
+      settings: {
+        // Should be sufficiently larger than expected job runtime
+        lockDuration: STATE_MONITORING_QUEUE_MAX_JOB_RUNTIME_MS,
+        // We never want to re-process stalled jobs
+        maxStalledCount: 0
+      }
+    })
+  }
+
+  registerQueueEventHandlers(queue) {
+    // Add handlers for logging
+    queue.on('global:waiting', (jobId) => {
+      this.log(`Queue Job Waiting - ID ${jobId}`)
+    })
+    queue.on('global:active', (jobId, jobPromise) => {
+      this.log(`Queue Job Active - ID ${jobId}`)
+    })
+    queue.on('global:lock-extension-failed', (jobId, err) => {
+      this.logError(
+        `Queue Job Lock Extension Failed - ID ${jobId} - Error ${err}`
+      )
+    })
+    queue.on('global:stalled', (jobId) => {
+      this.logError(`stateMachineQueue Job Stalled - ID ${jobId}`)
+    })
+    queue.on('global:error', (error) => {
+      this.logError(`Queue Job Error - ${error}. Queuing another job...`)
+    })
+
+    // Re-queue a new job when the current job fails or succeeds
+    queue.on('completed', (job, result) => {
+      this.log(
+        `Queue Job Completed - ID ${job?.id} - Result ${JSON.stringify(
+          result
+        )}. Queuing another job...`
+      )
+      if (result?.jobFailed) {
+        this.enqueueJobAfterSuccess(queue, job, result)
+      } else {
+        this.enqueueJobAfterFailure(queue, job)
+      }
```

Comment:
these event handlers are mostly the same as before. major difference is that subsequent jobs rely on previous job data and results instead of class variables. this eliminates state entirely and means you can debug a job purely based on its input (jobData) and output (result), which are both easy to read in bull-board (`/health/bull` endpoint)

Reply:
am generally fine with this, but am slightly concerned that new job creation is a bit hidden, like you wouldn't think to look at the eventHandler function for job production right

offhand, not sure of a better solution, but maybe think about it and we can discuss

Reply:
changed to the way we talked about where the event registration function takes success and failure callbacks